python/link_prediction.py

```diff
@@ -0,0 +1,771 @@
+
+    # Device type handling
+    if link_prediction_parameters.device_type == CUDA_DEVICE and torch.cuda.is_available() is True:
+        link_prediction_parameters.device_type = CUDA_DEVICE
+    else:
+        link_prediction_parameters.device_type = CPU_DEVICE
+
+    # Lists handling=generator expression + unpacking
+    if type(link_prediction_parameters.hidden_features_size) == tuple:
+        link_prediction_parameters.hidden_features_size = [*(x for x in link_prediction_parameters.hidden_features_size)]
+
+    if type(link_prediction_parameters.attn_num_heads) == tuple:
+        link_prediction_parameters.attn_num_heads = [*(x for x in link_prediction_parameters.attn_num_heads)]
+
+    print("END")
+    print(link_prediction_parameters)
+
+    return mgp.Record(status=1, message="OK")
+
+
+@mgp.read_proc
+def train(
+    ctx: mgp.ProcCtx,
+) -> mgp.Record(training_results=mgp.Any, validation_results=mgp.Any):
+    """Train method is used for training the module on the dataset provided with ctx. By taking decision to split the dataset here and not in the separate method, it is impossible to retrain the same model.
+
+    Args:
+        ctx (mgp.ProcCtx, optional): Reference to the process execution.
+
+    Returns:
+        mgp.Record: It returns performance metrics obtained during the training on the training and validation dataset.
+    """
+    # Get global context
+    global training_results, validation_results, predictor, model, graph, new_to_old, old_to_new
+
+    # Reset parameters of the old training
+    _reset_train_predict_parameters()
+
+    # Get some data
+    graph, new_to_old, old_to_new = _get_dgl_graph_data(ctx)  # dgl representation of the graph and dict new to old index
+
+    # Check if there are no edges in the dataset, assume that it cannot learn effectively without edges. E2E handling.
+    if graph.number_of_edges() == 0:
+        raise Exception("No edges in the dataset. ")
+
+    # TEST: Currently disabled
+    _conversion_to_dgl_test(graph=graph, new_to_old=new_to_old, ctx=ctx, node_features_property=link_prediction_parameters.node_features_property)
+
+    """ Train g is a graph which has removed test edges. Others are positive and negative train and test graphs
+    """
+
+    # Insert in the hidden_features_size structure if needed
+    _load_feature_size(graph.ndata[link_prediction_parameters.node_features_property].shape[1])
+
+    # Split the data
+    train_g, train_pos_g, train_neg_g, val_pos_g, val_neg_g = preprocess(graph, link_prediction_parameters.split_ratio)
+
+    # Create a model
+    model = create_model(
+        layer_type=link_prediction_parameters.layer_type,
+        hidden_features_size=link_prediction_parameters.hidden_features_size,
+        aggregator=link_prediction_parameters.aggregator,
+        attn_num_heads=link_prediction_parameters.attn_num_heads,
+    )
+    # Create a predictor
+    predictor_hidden_size = link_prediction_parameters.hidden_features_size[-1]
```

Comment:
It could be implemented within the model, but I find it more clear when the model and the predictor are separated. And in this way, there is a clear distinction between the node-embedding task and the link-prediction task which means that in the future node-embeddings can be potentially used for more tasks. Gradient backpropagation can be also more easily debugged.